include/cusp/detail/device/detail/coo.h

```diff
@@ -61,37 +61,37 @@ void coo_elementwise_transform_simple(const Matrix1& A,
     cusp::array1d<IndexType,MemorySpace> cols(A_nnz + B_nnz);
     cusp::array1d<ValueType,MemorySpace> vals(A_nnz + B_nnz);
 
-    thrust::copy(A.row_indices.begin(),    A.row_indices.end(),    rows.begin());
-    thrust::copy(B.row_indices.begin(),    B.row_indices.end(),    rows.begin() + A_nnz);
-    thrust::copy(A.column_indices.begin(), A.column_indices.end(), cols.begin());
-    thrust::copy(B.column_indices.begin(), B.column_indices.end(), cols.begin() + A_nnz);
-    thrust::copy(A.values.begin(),         A.values.end(),         vals.begin());
+    amgx::thrust::copy(A.row_indices.begin(),    A.row_indices.end(),    rows.begin());
+    amgx::thrust::copy(B.row_indices.begin(),    B.row_indices.end(),    rows.begin() + A_nnz);
+    amgx::thrust::copy(A.column_indices.begin(), A.column_indices.end(), cols.begin());
+    amgx::thrust::copy(B.column_indices.begin(), B.column_indices.end(), cols.begin() + A_nnz);
+    amgx::thrust::copy(A.values.begin(),         A.values.end(),         vals.begin());
 
     // apply transformation to B's values 
-    thrust::transform(B.values.begin(), B.values.end(), vals.begin() + A_nnz, op);
+    thrust_wrapper::transform(B.values.begin(), B.values.end(), vals.begin() + A_nnz, op);
```

Comment:
Is there semantic difference between `amgx::thrust` and `thrust_wrapper` - in which situations which is called? Maybe possible to merge them?
On a related note - do those support `thrust::device.on(stream)`? I think this is something that will be useful at some point.

Reply:
So, `amgx::thrust` is just standard thrust, but with the `amgx` namespace added because thrust has a very severe limitation that means two libraries using thrust will clash on some static state. This has affected majors users of AmgX due to many of them also using thrust in their code base.

The reason there is also `thrust_wrapper` is that I am slowly trying to remove all of the implicit synchronisation that was introduced by thrust a few years ago. Overnight we suddenly had a synchronisation after every thrust call, which impacts performance in the latency sensitive parts. I'm trying to slowly undo this, but I don't have the time to wrap every single site and test right now, so I have done as much as I can.

> On a related note - do those support thrust::device.on(stream)? I think this is something that will be useful at some point.

The purpose of `thrust_wrapper` is to make it easier to introduce changes like this. Putting all of the thrust calls on a stream will be straightforward, and we can also add a template to override for individual calls too, as necessary.